add(raw  RFC): 34/WAKU2-PEER-EXCHANGE

kaiserd opened the pull request:
This PR addresses #495 with a raw RFC for a simple peer exchange protocol.

## possible variations

* This protocol neither allows the requester to provide an indication for the desired  discovery method, e.g. discv5.
Nor does it allow the responder to include the used method in the response. It simply suggests using discv5.
We could add such indications. Imo, this would add little value at the cost of complexity.

* The response could be changed to contain `ENRs` instead of `multiaddrs`. Imo, multiaddrs are better suited as requesters are not expected to engage in discv5.

* We could remove the request message including the `numPeers`  and simply use opening the connection with the respective protocol ID (`/vac/waku/peer-exchange/2.0.0-alpha1`) as the request. (Similar to what `identify` does.)

cc @Menduist

kaiserd committed:
add: first raw version of peer exchange RFC

Menduist commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,123 @@
+
+The peer exchange protocol specified in this document is a simple request-response protocol.
+As Figure 1 illustrates, the requesting node sends a request to a peer, which acts as the responder.
+The responder replies with a list of [multiaddresses](https://docs.libp2p.io/concepts/addressing/).
```
Ideally we would use a Signed Peer Record, but I understand this wouldn't be compatible with discv5

But we should at least include the PeerId (which can be extracted from the ENR)

kaiserd replied:
I will look into this.
It should be in line with the [StrictNoSign Policy](https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/p2p-interface.md#why-are-we-using-the-strictnosign-signature-policy).
Also have to check discv5 ENRs regarding this.

D4nte replied:
[Waku ENR](https://rfc.vac.dev/spec/31/) payload is derived from [EIP-778](https://eips.ethereum.org/EIPS/eip-778).

It contains a `signature`of1 the content, signed by the public key encoded in the `secp256k1` field (signature is defined by identity scheme specified in `id` field).

Do we need a similar mechanism for this protocol? Should we simply return the ENRs to include this security feature?

[EIP-778](https://eips.ethereum.org/EIPS/eip-778) does not specify why the signature is provided. After all, if the details are incorrect then connection would just failed because remote peer id must back the public key in the ENR.

Menduist replied:
> StrictNoSign Policy.

This only applies to gossipsub to avoid linking a message to it's originator
Here, a peer is expected to announce him himself on the network, so we already know he is the originator of this data

kaiserd replied:
Yes. I was looking for implications I might have missed :sweat_smile: . Did not find any problems so far.
I'm switching to ENRs in the response.

I'll add your statement to the security section to make that explicit.

kaiserd replied:
Still, ENRs could be used to track peers over multiple "sessions", and maybe used to facilitate some linking attacks.
Ill mention that in security section. I'll look more into that when I continue with the anonymity analysis.

Menduist replied:
Yes, though this is more a discv5 issue than a PX issue

See https://github.com/status-im/nim-libp2p/issues/727 for PeerId (which is the identity used in the ENR) rotation discussion
(TLDR: we should change peerid when we change IP, which is not trivial to do)

kaiserd replied:
Yes. I remeber the discord convo :).
I touched on possible implications in the security section.

Menduist commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,123 @@
+
+In order to protect its anonymity, the responder MUST NOT provide peers from its actively used peer list as this opens pathways to *Neighbourhood Surveillance* attacks, as described in the
+[Security/Privacy Considerations Section](#securityprivacy-considerations).
+The responder SHOULD retrieve a new random set of peers via [33/WAKU2-DISCV5](https://rfc.vac.dev/spec/33/).
```
Kind of an implementation detail, but do you foresee this as being "a DISCV5 request for every incoming PX request", or "a DISCV5 request every X minutes / whatever, and for this duration I reply the same set"?

kaiserd replied:
At the moment, I wanted to leave that open.
I should discuss this in the "implantation suggestions" section.

Its a trade-off between anonymity and efficiency.
Sending the same set of peers to various nodes opens pathways to neighbourhood surveillance.

D4nte replied:
For efficiency purposes, especially in the context of a web app loading, it would be better if the responder retrieves and cache a random set of peers via discv5 first and then deliver some of those peers in the response I suggest we strongly recommended the .

If the responder triggers discv5 upon receiving a request, I am afraid the latency would be too high to be usable in a web app context.

I can see this is also mentioned in the Security section.

I think we should rephrase this to discourage discv5 being trigger upon request and encourage the usage of a cache.

kaiserd replied:
Addressed in 33263c3 while protecting the anonymity properties. I did not thoroughly analyse the implications of reusing peers in response sets yet. But the phrasing now makes caching a SHOULD and reusing exchange peers a MAY.

Menduist commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,123 @@
+Heavily used static nodes also add a centralized element. Downtime of such a node might significantly impact the network.
+
+However, the resource efficiency of this protocol comes at an anonymity cost, which is explained in the [Security/Privacy Considerations](#securityprivacy-considerations) section.
+This protocol SHOULD only be used if [33/WAKU2-DISCV5](https://rfc.vac.dev/spec/33/) is infeasible.
```
Not sure how much of an issue this is, but doing this exposes "resource restricted devices" in the network

kaiserd replied:
Yes. I should mention that :).

kaiserd committed:
add: protocol identifier

kaiserd committed:
add figure

kaiserd committed:
fix

Menduist reviewed (commented):
Nice first draft!

So, my understanding of this PR is that it's more a "delegated DHT query" than a "Generalized Peer Exchange", right?

If so, maybe we could embrace it further, by making it an actual "Forwarding DHT mechanism", that could also be used for instance in codex for content discovery

As far as actual Peer Exchange, I've actually been thinking that Rendez-Vous could do the job with a few tweaks (being enabled on all nodes, don't send peers you're currently connected to, shuffle your responses)

kaiserd committed:
fix

kaiserd commented:
Thank you for your feedback!

> So, my understanding of this PR is that it's more a "delegated DHT query" than a "Generalized Peer Exchange", right?

Yes. The purpose is to have a simple peer exchange protocol that allows resource restricted nodes to get a set of (preferably randomly sampled) peers.

> If so, maybe we could embrace it further, by making it an actual "Forwarding DHT mechanism", that could also be used for instance in codex for content discovery.

For now, we wanted to keep it simple and implement/ship something soon.
We could add functionality though. What would expect from an actual "Forwarding DHT mechanism"?

> As far as actual Peer Exchange, I've actually been thinking that Rendez-Vous could do the job [...]

I thought about that, too. My thought process was: the simple peer-exchange would be simpler :sweat_smile:  , and come with fewer -- or at least better understood -- security implications.
Also, by the virtue of bridging discv5, the simple peer-exchange protocol can help achieve a better distribution.
Maybe, we can discuss this further. I will also look into this more.

D4nte reviewed (commented):
Looks good, please address the comment regarding caching.

kaiserd committed:
address feedback; protect anonymity properties

kaiserd commented:
Thank you for the feedback :).
Caching addressed in 33263c3. See [comment](https://github.com/vacp2p/rfc/pull/515#discussion_r931323886).

I will look more into the implication of ENR/PeerInfo signing, to not violate the noSignPolicy.

> Do we need a similar mechanism for this protocol? Should we simply return the ENRs to include this security feature?

If the outcome is we should sign, I'd suggest returning ENRs. In this first raw version, I used just multiaddrs to avoid signing.

kaiserd committed:
chache size recommendation

kaiserd commented:
> you assume numPeers to be 6. I am guessing you are getting this number from the recommended gossipsub parameters D=6?

Yes. I clarified that and moved it into the "implementation suggestions section" while keeping the "theory/semantics" sections more general".
Updated  in: eb952883

> While we want to make it possible for light client to use Waku Relay, it is not scalable at the moment so we are more likely to push for light push + filter,
especially once we improve filter and until we have an available strategy to make relay scalable (wip, action item on my side).

> All that to say that I don't think we should use 6. I hope that a service node can serve in the order of thousands light clients...

I added that the cache size also depends on the number of peers that are expected to be served.

> Maybe for now we do not provide any recommendation on caching size?

I mentioned, that we will provide a better strategy in the future.
The recommendation is now in the "suggestions" sections, which is not normative.
Imo, the suggestion helps understanding what points there are to consider regarding cache size.
wdyt?

kaiserd committed:
suggestions section -> only recommendations

remove SHOULD

kaiserd committed:
multiaddrs -> ENRs

kaiserd committed:
fixes and editing

kaiserd committed:
proto format

kaiserd committed:
editing

kaiserd commented:
I updated the RFC using [ENRs](https://rfc.vac.dev/spec/31/) in the response. Also some editing.

With this, I covered the feedback on PeerInfo / ENR, and caching.
wdyt?

The ENR is still wrapped into a `PeerInfo` which would allow adding items later. Since this does not add much complexity, nor does it add more transmitted data, Imo, this is OK.

Wdyt about the 3rd variation in the PR description? Should we keep the request message containing `numPeers`?.

D4nte commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,172 @@
+
+# Background and Motivation
+
+Resource restricted devices may not feasibly be able to take part in distributed random sampling ambient peer discovery protocols such as [33/WAKU2-DISCV5](https://rfc.vac.dev/spec/33/).
```
```suggestion
It may not be feasible on resource restricted devices to take part in distributed random sampling ambient peer discovery protocols such as [33/WAKU2-DISCV5](https://rfc.vac.dev/spec/33/).
```

D4nte commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,172 @@
+# Background and Motivation
+
+Resource restricted devices may not feasibly be able to take part in distributed random sampling ambient peer discovery protocols such as [33/WAKU2-DISCV5](https://rfc.vac.dev/spec/33/).
+The Waku peer discovery protocol specified in this document allows resource restricted devices to request a list of peers from a (stronger) peer.
```
```suggestion
The Waku peer discovery protocol specified in this document allows resource restricted devices to request a list of peers from a service node.
```
I believe this is the terminology we want to use.

D4nte commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,172 @@
+
+Resource restricted devices may not feasibly be able to take part in distributed random sampling ambient peer discovery protocols such as [33/WAKU2-DISCV5](https://rfc.vac.dev/spec/33/).
+The Waku peer discovery protocol specified in this document allows resource restricted devices to request a list of peers from a (stronger) peer.
+Network parameters necessary to connect to this stronger peer can, for instance, be learned from a static bootstrapping method.
```
```suggestion
Network parameters necessary to connect to this service node COULD be learned from a static bootstrapping method or using [EIP-1459: Node Discovery via DNS](https://eips.ethereum.org/EIPS/eip-1459).
```

Just a suggestion.

D4nte commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,172 @@
+Resource restricted devices may not feasibly be able to take part in distributed random sampling ambient peer discovery protocols such as [33/WAKU2-DISCV5](https://rfc.vac.dev/spec/33/).
+The Waku peer discovery protocol specified in this document allows resource restricted devices to request a list of peers from a (stronger) peer.
+Network parameters necessary to connect to this stronger peer can, for instance, be learned from a static bootstrapping method.
+The advantage of using Waku peer exchange to discover new peers over simply adding static bootstrap peers to the peer list is a more even load distribution.
```
```suggestion
The advantage of using Waku peer exchange to discover new peers over using a static list peer or dns discovery is a more even load distribution.
```

kaiserd replied:
I used:
`using a static peer list or DNS discovery is a more even load distribution`
e4a4728

D4nte commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,172 @@
+The Waku peer discovery protocol specified in this document allows resource restricted devices to request a list of peers from a (stronger) peer.
+Network parameters necessary to connect to this stronger peer can, for instance, be learned from a static bootstrapping method.
+The advantage of using Waku peer exchange to discover new peers over simply adding static bootstrap peers to the peer list is a more even load distribution.
+If a lot of (resource restricted) nodes would use static bootstrap nodes as relay or store nodes, the load on these would be very high.
```
```suggestion
If a lot of (resource restricted) nodes would use the same service nodes as relay or store nodes, the load on these would be very high.
```

D4nte commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,172 @@
+
+# Background and Motivation
+
+Resource restricted devices may not feasibly be able to take part in distributed random sampling ambient peer discovery protocols such as [33/WAKU2-DISCV5](https://rfc.vac.dev/spec/33/).
```
Relative link are usually better `/spec/33/` as it makes it easier to migrate the website in the future (not that I expect any migration)

kaiserd replied:
Ill do that after applying the rest of the feedback.

kaiserd replied:
Done in 99e7782f1

D4nte commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,172 @@
+The responder replies with a list of ENRs as specified in [31/WAKU2-ENR](https://rfc.vac.dev/spec/31/).
+The [multiaddresses](https://docs.libp2p.io/concepts/addressing/) used to connect to the respective peers can be extracted from the ENRs.
+
+![Figure 1: The responder provides a list of ENRs to the requester. These ENRs contain the information necessary for connecting to the respective peers.](../../../../static/rfcs/34/protocol.svg)
```
```suggestion
![Figure 1: The responder provides a list of ENRs to the requester. These ENRs contain the information necessary for connecting to the respective peers.](/rfcs/34/protocol.svg)
```
This should work.

D4nte commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,172 @@
+This both protects the responder's anonymity as well as helps distributing load.
+
+To allow for fast responses, responders SHOULD retrieve peers unsolicited (before receiving a query)
+and maintain a queue of peers for the purpose of providing them in peer-exchange responses.
```
```suggestion
and maintain a queue of peers for the purpose of providing them in peer exchange responses.
```

D4nte commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,172 @@
+Requesters send a simple message request which causes responders to engage in ambient peer discovery to retrieve a new random peer set.
+As a mitigation, responders MAY feature a `seen cache` for requests and only answer once per time interval.
+The exchange-peer cache discussed in [Theory and Protocol Semantics Section](#theory-and-protocol-semantics) also provides mitigation.
+Still, fequent queries can tigger the refresh cycle more often. The `seen cache` MAY be used in conjunction to provide additional mitigation.
```
```suggestion
Still, frequent queries can tigger the refresh cycle more often. The `seen cache` MAY be used in conjunction to provide additional mitigation.
```

D4nte reviewed (approved):
LGTM.

I propose to mention DNS discovery too, feel free not to merge it.
Couple of typos and suggestion re hrefs.

oskarth reviewed (approved):
LGTM as a raw spec, nice to see this coming along! I like the anonymity preservation technique and the simplicity.

kaiserd committed:
Update content/docs/rfcs/34/README.md

Co-authored-by: Franck R. <franck@status.im>

kaiserd committed:
Update content/docs/rfcs/34/README.md

Co-authored-by: Franck R. <franck@status.im>

kaiserd committed:
Update content/docs/rfcs/34/README.md

Co-authored-by: Franck R. <franck@status.im>

kaiserd committed:
feedback

kaiserd committed:
Update content/docs/rfcs/34/README.md

Co-authored-by: Franck R. <franck@status.im>

kaiserd committed:
Update content/docs/rfcs/34/README.md

Co-authored-by: Franck R. <franck@status.im>

kaiserd committed:
Update content/docs/rfcs/34/README.md

Co-authored-by: Franck R. <franck@status.im>

kaiserd committed:
relative links

kaiserd commented:
> LGTM.
> 
> I propose to mention DNS discovery too, feel free not too. Couple of typos and suggestion re hrefs.

Thank you for the feedback. I applied it all :).

kaiserd committed:
update ENR anonymity implications

kaiserd commented:
I updated the ENR anonymity implications in 3808217. See this [comment](https://github.com/vacp2p/rfc/pull/515).
Wdyt @Menduist @D4nte ?
If you are OK with it, I'd merge.

kaiserd committed:
add spec to index

kaiserd committed:
Update content/docs/rfcs/34/README.md

Co-authored-by: Franck R. <franck@status.im>

fryorcraken commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,171 @@
+This document provides recommended choices for the cache size in the [Implementation Suggestions Section](#implementation-suggestions).
+
+Requesters, in the context of the specified peer exchange protocol, SHOULD be resource restricted devices.
+While any node could technically act as a requester, using the peer exchange protocol comes with two drawbacks
```
```suggestion
While any node could technically act as a requester, using the peer exchange protocol comes with two drawbacks:
```

fryorcraken commented on content/docs/rfcs/34/README.md:
```diff
@@ -0,0 +1,171 @@
+
+## Discovery Interface
+
+Implementations can implement the [libp2p discovery interface](https://github.com/status-im/nim-libp2p/issues/140).
```
```suggestion
Implementations can implement the libp2p discovery interface (e.g. [nim](https://github.com/status-im/nim-libp2p/issues/140), [javascript](https://github.com/libp2p/js-libp2p-interfaces/tree/master/packages/interface-peer-discovery)).
```

fryorcraken reviewed (approved):
LGTM

kaiserd committed:
Update content/docs/rfcs/34/README.md

Co-authored-by: fryorcraken.eth <110212804+fryorcraken@users.noreply.github.com>

kaiserd committed:
Update content/docs/rfcs/34/README.md

Co-authored-by: fryorcraken.eth <110212804+fryorcraken@users.noreply.github.com>
